geps/gep-1651.md

```diff
@@ -0,0 +1,268 @@
+
+### Adding `reachabiliy` attribute to GatewayClass
+
+See [Prior Art - Multiple Gateways Classes](#multiple-gateway-classes)
+
+## Survey of Prior Art
+
+These alternatives are a survey of existing approaches to support cluster
+local Gateways. Most are implementation specific and are not portable.
+
+### Special annotation/label
+
+Istio let's you specify an annotation `networking.istio.io/service-type` to
+change the underlying Kubernetes Service type to make it a ClusterIP type.
+
+### Re-use of AddressType Hostname
+
+Istio let's you re-use existing Gateway deployments by setting the address
+type to `Hostname` and the value to the Istio ingress Kubernetes Service. If an
+operator configures the Istio deployment to support cluster local traffic a
+Gateway implementation can select it using the `HostName` attribute.
+
+### Multiple Gateway Classes
+
+Some implementations support multiple deployments on a single cluster where each maps to a
+GatewayClass. One of these deployments can be configured to serve cluster local traffic. This is
+sub-optimal because this is implementation specific and the end-user is effectively managing the
+deployments themselves rather than infrastructure being automatically provisioned.
+
+Likewise, infrastructure providers may provide a fixed set of GatewayClasses with unique and fixed
+reachability. Thus GatewayClass name is a viable option to control reachability. But it is not
+flexible enough for implementations that can configure reachability when creating Gateways.
+Otherwise implementations would be forced to create unnecessary GatewayClasses that may cause
+redundant infrastructure to be provisioned.
+
+Additionally, if more attributes are added to GatewayClass to constrain Gateways in some
+form this leads to a combinatorial number of GatewayClassNames. For example, `foo-public` and
+`foo-cluster` are two GatewayClasses surfacing the values of a single attribute `reachability`.
+Let's say we want to enforce address types to just IP then our `gatewayClassName` would be:
+- `foo-public-ipv4`
+- `foo-public-ipv6`
+- `foo-cluster-ipv4`
+- `foo-cluster-ipv6`
+
+As howardjohn mentioned [here](https://github.com/kubernetes-sigs/gateway-api/pull/1653#issuecomment-1429992160):
+> having the ability to configure things at a higher level seems nice for Gateway, but being able to configure them on a per-Gateway basis remains important.
+
+## References
+
+- Knative - Private Services - https://knative.dev/docs/serving/services/private-services/#configuring-private-services
+- Initial Gateway GitHub Discussion - https://github.com/kubernetes-sigs/gateway-api/discussions/1247
+- Istio Support for Private Gateways - https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#automated-deployment
+
```

Comment:
another data point while collecting the state of the world
EnvoyGateway defines the service type of the gateway infrastructure -  `ClusterIP`/`Loadbalancer` at the `GatewayClass` level (linked via the `parametersRef` field)
https://github.com/envoyproxy/gateway/blob/cd18c968dc74c760770975a9d2e658217907c682/api/config/v1alpha1/envoyproxy_types.go#L95

Reply:
thanks - added